Add ERC: Flash Loans (#263)
ethereum/ERCs: ERCS/erc-7399.md

---
eip: 7399

> title: Flash Loans

Reviewer: @sambacha: Flexible Flash Loans?

Basically the difference between ERC3156 and ERC3799 is that the latter is a lot more flexible (giving power to users in exchange of some complexity).

> description: Interfaces and processes for flash loans

Reviewer: @sambacha, maybe we just copy the abstract from erc7399?

```
A flash loan is a smart contract transaction in which a lender smart contract lends assets to a borrower smart contract with the condition that the assets are returned, plus an optional fee, before the end of the transaction. This ERC specifies interfaces for lenders to accept flash loan requests, and for borrowers to take temporary control of the transaction within the lender execution. The process for the safe execution of flash loans is also specified.
```

> description: Interfaces and processes for flash loans

Reviewer 2: Your description should explicitly say what makes your proposal different from existing flash loans. I'm not familiar enough with DeFi to give a useful suggestion.

discussions-to: https://ethereum-magicians.org/t/erc7400-flash-loans/15211
status: Draft
type: Standards Track
category: ERC
created: 2023-07-25
requires: 20
---

## Abstract

A flash loan is a smart contract transaction in which a lender smart contract lends assets to a borrower smart contract with the condition that the assets are returned, plus an optional fee, before the end of the transaction. This ERC specifies interfaces for lenders to accept flash loan requests, and for borrowers to take temporary control of the transaction within the lender execution. The process for the safe execution of flash loans is also specified.

## Motivation

Early adopters of the flash loan pattern have produced different interfaces and different use patterns. The diversification is expected to intensify, and with it the technical debt required to integrate with diverse flash lending patterns.

Some of the high level differences in the approaches across the protocols include:

 - Syntax of the call initiating the flash loan.
 - Possibility of the loan receiver and the callback receiver being different addresses.

>  - Repayment approaches at the end of the transaction, where some pull the principal plus the fee from the loan receiver, and others where the loan receiver needs to manually return the principal and the fee to the lender.

Reviewer: @sambacha, given that we introduced this bullet point list as "Some of the high level differences in the approaches across the protocols include:"

Maybe we can say something like:

"Flash loan repayment mechanism, which can be pull (take the repayment from the borrower) or push (expect the borrower to send the repayment to the lender)".

 - Some lenders allow to flash mint any amount of their native asset without charging a fee, effectively allowing flash loans bounded by computational constraints instead of asset ownership constraints.

This ERC aims to consolidate all flash lenders into a common standard of maximum flexibility, so that borrowers can easily choose flash lenders without changing their code.

## Specification

> 

Reviewer 2: ```suggestion

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

```

This `amount` plus a `fee` defined by the `lender` in the same `asset` must be repaid before the end of `flash` call at a `repayment receiver` address defined by the `lender`.

The `flash` function is called by the `initiator`, who defines the `loan receiver`, the `callback receiver`, the `callback function`, the `asset` and the `amount`.

When the `initiator` calls `flash` in a `lender`. The `lender` will then transfer the `amount` of `asset` to the `loan receiver`.

The `lender`, after transferring `amount` of `asset` to the `loan receiver`, will execute the `callback function` on the `callback receiver`. The `lender` will include in this `callback function` call a number of parameters related to the loan as defined in this standard. 

 The `amount` and `fee` need to be transferred to a `repayment receiver` before the end of the `flash` call. The `fee` can be set to zero `asset`.

The `callback function` can return any arbitrary data which will be received by the `initiator` as the return value of the `flash` call.

### Lender Specification

A `lender` MUST implement the [ERC-7399](./eip-7399.md) interface.

```solidity

> // SPDX-License-Identifier: CC0

Reviewer 2: ```suggestion
// SPDX-License-Identifier: CC0-1.0
```

A `payment receiver` allows for the same flexibility on repayments as in borrows. Control flow and asset flow are independent.

The `amount` will be required in the callback function, which the `lender` took as a parameter. An alternative implementation which would embed the `amount` in the `data` parameter by the caller would require an additional mechanism for the receiver to verify its accuracy, and is not advisable.

A `fee` will often be calculated in the callback function, which the callback receiver must be aware of for repayment. Passing the `fee` as a parameter instead of appended to `data` is simple and effective.

Arbitrary callback functions on callback receivers allows to implement different behaviours to flash loans on callback receivers without the need for encoding a function router using the `data` argument. A function call type is 24 bytes of which the first 20 bytes are the target address and the last 4 bytes are the function signature.

> The `amount + fee` are pushed to the `payment receiver` to allow for the segregation of asset and control flows. While a "pull" architecture is more prevalent, "push" architectures are also common. For those cases where the `lender` can't implement a "push" architecture, a simple wrapper contract can offer an ERC-7399 external interface, while using liquidity from the `lender` using a "pull" architecture.

Reviewer 2: ```suggestion
The `amount + fee` are pushed to the `payment receiver` to allow for the segregation of asset and control flows. While a "pull" architecture is more prevalent, "push" architectures are also common. For those cases where the `lender` can't implement a "push" architecture, a simple wrapper contract can offer this proposal's external interface, while using liquidity from the `lender` using a "pull" architecture.
```

## Backwards Compatibility

> This EIP is a successor of [ERC-3156](./eip-3156.md). While not directly backwards compatible, a wrapper contract offering a ERC-7399 external interface with liquidity obtained from an ERC-3156 flash `lender` is trivial to implement.

Reviewer 2: ```suggestion
This EIP is a successor of [ERC-3156](./eip-3156.md). While not directly backwards compatible, a wrapper contract offering this proposal's external interface with liquidity obtained from an ERC-3156 flash `lender` is trivial to implement.
```
